manager/volume.go

```diff
@@ -614,7 +617,58 @@ func (m *VolumeManager) Expand(volumeName string, size int64) (v *longhorn.Volum
 	}
 
 	if v.Spec.Size > size {
-		return nil, fmt.Errorf("cannot expand volume %v with current size %v to a smaller size %v", v.Name, v.Spec.Size, size)
+		return nil, fmt.Errorf("longhorn volume %v spec size %v is greater than requested Persistent Volume Claim size %v", v.Name, v.Spec.Size, size)
+	}
+
+	kubernetesStatus := &v.Status.KubernetesStatus
+	if kubernetesStatus.PVCName != "" && kubernetesStatus.LastPVCRefAt == "" {
+		pvc, err := m.ds.GetPersistentVolumeClaim(kubernetesStatus.Namespace, kubernetesStatus.PVCName)
+		if err != nil {
+			return nil, err
+		}
+
+		requestedSize := resource.MustParse(strconv.FormatInt(size, 10))
+
+		// TODO: Should check for pvc.Spec.Resources.Requests.Storage() here, once upgrade API to v0.18.x.
+		pvcSpecValue, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		if !ok {
+			return nil, fmt.Errorf("cannot get request storage")
+		}
+		// TODO: Should check for pvc.Status.CapacityStorage() here, once upgrade API to v0.18.x.
+		pvcStatusValue, ok := pvc.Status.Capacity[corev1.ResourceStorage]
+		if !ok {
+			return nil, fmt.Errorf("cannot get capacity storage")
+		}
+
+		if pvcSpecValue.Cmp(requestedSize) == 0 && pvcStatusValue.Cmp(requestedSize) == 0 && v.Spec.Size == size {
+			// lhv size does not reflect pvc/pv size
+			logrus.Infof("Persistent Volume Claim %v expansion is not necessary since current size %v == %v", v.Name, v.Spec.Size, size)
+			return v, nil
+		} else if pvcSpecValue.Cmp(requestedSize) > 0 {
+			// this is possible due to user can patch lvh, but not patch pv and pvc
+			return nil, fmt.Errorf("persistent volume claim %v spec size %v is greater than requested persistent volume claim size %v", kubernetesStatus.PVCName, pvcSpecValue.Value(), requestedSize.Value())
+		} else if pvcStatusValue.Cmp(requestedSize) > 0 {
+			// this is possible due to user manually created pvc
+			return nil, fmt.Errorf("persistent volume claim %v status size %v is greater than requested persistent volume claim size %v", kubernetesStatus.PVCName, pvcStatusValue.Value(), requestedSize.Value())
+		} else if pvcSpecValue.Cmp(requestedSize) < 0 {
+			// only do pvc expand if requested size is larger
+			pvc.Spec.Resources = corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: requestedSize,
+				},
+			}
+
+			logrus.Infof("Persistent Volume Claim %v expansion from %v to %v requested", v.Name, v.Spec.Size, size)
+			_, err = m.ds.UpdatePersistentVolumeClaim(kubernetesStatus.Namespace, pvc)
+			if err != nil {
+				return nil, err
+			}
+
+			// return and CSI plugin call this API later for Longhorn volume expansion
+			return v, nil
+		}
+		// all other case belong to the CSI plugin call
+		logrus.Infof("CSI plugin call to expand volume %v", v.Name)
```

Comment:
Because you already tackle in PVC case, you should return here and the rest of the part will be tackled by the CSI plugin. Correct?